collection_json.py

```diff
@@ -6,6 +6,59 @@
 __version__ = '0.1.0'
 
 
+class ArrayProperty(object):
+
+    """Property which converts from any enumerable to a typed Array instance."""
+
+    def __init__(self, cls, name):
+        """Constructs typed array property
+
+        :param cls type: the type of objects expected in the array
+        :param name str: name of the property
+        """
+        self.cls = cls
+        self.name = name
+        self.value_attr = "_" + name
```

Comment:
So I was excited to implement this when I saw it, as it looks like a nice solution to the problem. The post even describes using WeakRefDictionary, addressing my immediate concerns about memory leaks via cycle. 

Unfortunately I don't think the implementation would be sound in our case. For dictionaries, you need to implement __hash__ and __eq__ and preserve the property if eq(o1) == eq(o2), then hash(o1) == hash(o2). Since we override eq and ne, we must implement a decent non-trivial hash function. 

The bigger problem is that mutability would cause the hash in this function to change, meaning the next look-up could fail. Worse yet, if users created multiple objects with the same values such they were equal, it would also cause the instances to hash to the same value and change the property of both objects -- a consequence of aggregating the objects together at the class level. I think this probably wouldn't work in all but the most trivial cases.